### PR TITLE
[Enhancement] avoid sync `listPartitionNames` when query iceberg & mv

### PR DESCRIPTION
**Pull request by dirtysalt**
## Why I'm doing:

`listPartitionNames` is a heavy operation on iceberg table. And in normal query process, we don't call that.

If we have MV, we will call that function in CBO stage. But we don't want waiting if there is no cache. 

We can just return a null(to tell CBO don't do MV rewrite) if there is no cache, and return cache value if cached.  So there is no sync operation during query.

## What I'm doing:

This PR is to add:
- `queryMVRerwrite` in connector metadata request context and connector traits
- set `queryMVRerwrite` to true when doing query MV rewrite
- during `listPartitionNames`, if query mv rewrite is true and there is no cache value, return null.
- add a session varibale `ENABLE_CONNECTOR_ASYNC_LIST_PARTITIONS` to enable this feature(off by default)

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java [hunk 1]
```diff
@@ -204,6 +205,24 @@ public Map<String, Partition> getPartitions(String dbName, String tableName, lon
...
+                                           ExecutorService executorService) {
+        // optimization for query mv rewrite, we can optionally return null to bypass it.
+        // if we don't have cache right now, which means it probably takes time to load it during query,
+        // so we can do load in background while return null to bypass this synchronous process.
+        if (requestContext.isQueryMVRewrite()) {
+            long snapshotId = requestContext.getSnapshotId();
+            IcebergTableName key = new IcebergTableName(dbName, tableName, snapshotId);
+            Map<String, Partition> cacheValue = partitionCache.getIfPresent(key);
+            if (cacheValue == null) {
+                partitionCache.refresh(key);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java:219]: are you sure this is async?

File: fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadatRequestContext.java [hunk 2]
```diff
@@ -25,5 +26,17 @@ public void setTableVersionRange(TableVersionRange value) {
     public TableVersionRange getTableVersionRange() {
         return tableVersionRange;
     }
+
+    public void setQueryMVRewrite(boolean v) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadatRequestContext.java:30]: do not introduce mv concepts to the Connector， I think `forceUseCache` is better

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java (hunk 1 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java:219]: it's async when there is value.
and it's sync when there is no value.

Damn, have to fix this.

**Review by Youngwb — DISMISSED** [review]

File: fe/fe-core/src/test/java/com/starrocks/connector/hive/ReplayMetadataMgr.java [hunk 3]
```diff
@@ -139,7 +140,8 @@ private List<Column> mockColumns(List<String> partitionColumns, List<String> dat
     }
 
     @Override
-    public List<String> listPartitionNames(String catalogName, String dbName, String tableName) {
+    public List<String> listPartitionNames(String catalogName, String dbName, String tableName,
```
[the remark this answers is not in the document]
Comment (dirtysalt) [reply fe/fe-core/src/test/java/com/starrocks/connector/hive/ReplayMetadataMgr.java:143]: I don't understand what you mean here.  MV rewrite is on by default.  I don't change any logic right here.

**Comment by eshishki** [comment]
users would expect their mv to work
we build mv for a reason, to accelerate slow queries

this patch will make performance and response time inconsistent
why you think it is better to execute original slow query than to wait for catalog response?

**Comment by dirtysalt** [comment]
> users would expect their mv to work we build mv for a reason, to accelerate slow queries
> 
> this patch will make performance and response time inconsistent why you think it is better to execute original slow query than to wait for catalog response?

@eshishki 

Because we have seen so many cases that 
- iceberg table is huge and takes a very longe time to get partitions.
- iceberg update speed is much faster than mv refresh speed. 

So everytime there is a mv & iceberg query,  at CBO stage,  it takes time to build partitions for mv rewrite, which adds latency.  Remember it fails to rewrite query to use MV, it still takes time to build partitions.

So there is a tradeoff between
a. wait for catalog response for partitions(T0) + MV execution time(T1)
b. wait for catalog response partitions(T0) + iceberg execution time(T2)
c. iceberg execution time(T2)

And we have seen many cases fall into (b) case.

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 4]
```diff
@@ -2250,6 +2249,9 @@ public SessionVariable setHiveTempStagingDir(String hiveTempStagingDir) {
     @VarAttr(name = CONNECTOR_INCREMENTAL_SCAN_RANGE_SIZE)
     private int connectorIncrementalScanRangeSize = 500;
 
+    @VarAttr(name = ENABLE_CONNECTOR_ASYNC_LIST_PARTITIONS)
+    private boolean enableConnectorAsyncListPartitions = false;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:2253]: why false default? when set to true?

**Review by Seaven — DISMISSED** [review]

**Review by stephen-shelby — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 4 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:2253]: @Youngwb I think we can enable this feature later.  And i've seen the above comments and he gets his point:  async partition does not work consistently with MV, which probably confuses user.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java [hunk 5]
```diff
@@ -198,12 +201,33 @@ public View getView(String dbName, String viewName) {
...
+            long snapshotId = requestContext.getSnapshotId();
+            IcebergTableName key =
+                    new IcebergTableName(icebergTable.getCatalogDBName(), icebergTable.getCatalogTableName(), snapshotId);
+            Map<String, Partition> cacheValue = partitionCache.getIfPresent(key);
+            if (cacheValue == null) {
+                backgroundExecutor.submit(() -> partitionCache.refresh(key));
+                return null;
+            }
+        }
+        return IcebergCatalog.super.listPartitionNames(icebergTable, requestContext, executorService);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java:228]: delegate.listPartitionNames

**Review by Youngwb — DISMISSED** [review]

**Comment by eshishki** [comment]
@dirtysalt i understand that it might be useful in some cases, but can we do a session variable for this or better yet configurable for mv?

that they i can make this tradeoff when i need to

**Comment by dirtysalt** [comment]
@eshishki  thanks for your proposal. And after second thought, I think you are right.

And I've added a session variable to control this behaviour

> add a session varibale `enable_connector_async_list_partitions` to enable this feature(off by default)

So by default it's false, and it's behaves like before. And if it's set to true, then it won't wait if parttion names not cached.


> @dirtysalt i understand that it might be useful in some cases, but can we do a session variable for this or better yet configurable for mv?
> 
> that way i can make this tradeoff when i need to

**Comment by dirtysalt** [comment]
@eshishki what you mean `per mv option`?  I don't quite understand how mv option can affect this process?

**Comment by eshishki** [comment]
@dirtysalt regular analyst user or some tableau dashboard might not know about this nuances, mv, and other optimisations and can't change session variable before each query

mv are usually created by dba or etl engineer who are best suited to make this tradeoff
i see now that in ALTER MATERIALIZED VIEW i can put session property and that is exactly what i would need

**Review by kangkaisen — APPROVED** [review]

**Comment by dirtysalt** [comment]
@eshishki 

I've merged this PR in advance. But I agree with your point.  this flag is better to be put as a mv property not a session variable.

**Pull request merged**